File: problem138.java
//problem 138 /copy list with random pointer
class Solution {
    public Node copyRandomList(Node head) {
         if (head == null) {
             return null;
         }
        copynext(head);
        copyrandom(head);
        return splitlist(head);
    }
    private void copynext(Node head) {
        while (head != null) {
            Node newnode = new Node(head.val,head.next,head.random);
            //newnode.random = head.random;
            //newnode.next = head.next;
            head.next = newnode;
            head = head.next.next;
        }
    }
    private void copyrandom(Node head) {
        while (head != null) {
            if (head.next.random != null) {
                head.next.random = head.random.next;
            }
            head = head.next.next;
        }
    }
    private Node splitlist(Node head) {
        Node newhead = head.next;
        while (head != null) {
            Node temp = head.next;
            head.next = temp.next;
            head = head.next;
            if (temp.next != null) {
                temp.next = temp.next.next;
            }
        }
        return newhead;
    }
}

File: problem138.py
#problem 138 / copy list with random pointer

# Definition for a Node.
class Node(object):
    def __init__(self, val, next, random):
        self.val = val
        self.next = next
        self.random = random

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        if not head:
            return head
        dummy = head
        while dummy:
            temp = Node(dummy.val,dummy.next,None)
            dummy.next = temp
            dummy = dummy.next.next
        dummy = head
        while dummy:
            if dummy.random:
                dummy.next.random = dummy.random.next
            else:
                dummy.next.random = None
            dummy = dummy.next.next
        oldlist = head
        newlist = head.next
        newhead = head.next
        while oldlist:
            oldlist.next = oldlist.next.next
            newlist.next = newlist.next.next if newlist.next else None
            oldlist = oldlist.next
            newlist = newlist.next
        return newhead